src/main.rs

```diff
@@ -39,42 +65,49 @@ fn distribute(
         parties[i].set_group_nonces(B.clone());
     }
 
-    total_compute_secret_time
+    (total_compute_secret_time, public_keys)
 }
 
 #[allow(non_snake_case)]
-fn select_parties<RNG: RngCore + CryptoRng>(N: usize, T: usize, rng: &mut RNG) -> Vec<usize> {
-    let mut indices: Vec<usize> = Vec::new();
+fn select_parties<RNG: RngCore + CryptoRng>(
+    key_owners: &Vec<usize>,
+    T: usize,
+    rng: &mut RNG,
+) -> HashMap<usize, HashSet<usize>> {
+    let mut signers = HashMap::new();
+    let mut pts = Vec::new();
 
-    for i in 0..N {
-        indices.push(i);
+    for i in 0..key_owners.len() {
+        pts.push(i);
     }
 
-    while indices.len() > T {
-        let i = rng.next_u64() as usize % indices.len();
-        indices.swap_remove(i);
+    while pts.len() > T {
+        let i = rng.next_u64() as usize % pts.len();
+        pts.swap_remove(i);
     }
 
-    indices
+    for k in pts {
+        let s = signers.entry(key_owners[k]).or_insert(HashSet::new());
+        s.insert(k);
+    }
+
+    signers
 }
 
 // There might be a slick one-liner for this?
 fn collect_signatures(
     parties: &Vec<Party>,
-    signers: &Vec<usize>,
+    signers: &HashMap<usize, HashSet<usize>>,
     nonce_ctr: usize,
     msg: &String,
 ) -> Vec<SignatureShare> {
-    let mut sigs = Vec::new();
-    for i in 0..signers.len() {
-        let party = &parties[signers[i]];
-        sigs.push(SignatureShare {
-            id: party.id.clone(),
-            z_i: party.sign(&msg, &signers, nonce_ctr),
-            public_key: party.public_key.clone(),
-        });
-    }
-    sigs
+    signers
+        .keys()
+        .map(|party_id| SignatureShare {
+            party_id: party_id.clone(),
```

Comment:
`party_id` is a reference to a usize, and should not need to be cloned. I assume the need here is to deference it.
```suggestion
            party_id: *party_id,
```

Another option would be to pattern match on it in line 106 as |&party_id|.